### PR TITLE
[mediastream] Add extractor

### LINKED ISSUE #4431: Add support for Winsports.co
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.07.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [x] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

Colombia

### Example URLs

https://www.winsports.co/videos/observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548
https://www.winsports.co/videos/siempre-castellanos-gran-atajada-del-portero-cardenal-para-evitar-la-caida-de-su-arco-60536

### Provide a description that is worded well enough to be understood

Yt-dlp can't download video or audio from the site 

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vF', 'https://www.winsports.co/videos/observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.07.18 [135f05e] (win32_exe)
[debug] Python 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 2022-07-14-git-882aac99d2-essentials_build-www.gyan.dev (setts), ffprobe 2022-07-14-git-882aac99d2-essentials_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] [generic] Extracting URL: https://www.winsports.co/videos/observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548
[generic] observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548: Downloading webpage
WARNING: [generic] Falling back on generic information extractor.
[generic] observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://www.winsports.co/videos/observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548
Traceback (most recent call last):
  File "yt_dlp\YoutubeDL.py", line 1431, in wrapper
  File "yt_dlp\YoutubeDL.py", line 1506, in __extract_info
  File "yt_dlp\extractor\common.py", line 644, in extract
  File "yt_dlp\extractor\generic.py", line 4108, in _real_extract
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.winsports.co/videos/observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548
```

### PR DESCRIPTION
**Pull request by elyse0**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Superseeds #5532

Extending the PR from HobbyistDev, this PR adds two new extractors: MediaStreamIE and WinSportsVideoIE. 

MediaStream is a generic video provider used in sites like WinSports, Multimedios, AmericaTv and CaracolTv.

Fixes #4431

I can't confirm #4425 is fixed because I get geo-blocked on all the videos

### WinSports extractor

```bash
[debug] Command-line config: ['https://www.winsports.co/videos/observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548', '--verbose', '-F']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.11.11 [8b644025b] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: d7b17cbb7
[debug] Python 3.10.7 (CPython x86_64 64bit) - Linux-5.19.0-23-generic-x86_64-with-glibc2.36 (OpenSSL 3.0.5 5 Jul 2022, glibc 2.36)
[debug] exe versions: ffmpeg N-108931-g4dda3b1653-20221104 (setts), ffprobe N-108931-g4dda3b1653-20221104, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, certifi-2022.09.24, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1730 extractors
[debug] [WinSportsVideo] Extracting URL: https://www.winsports.co/videos/observa-aqui-los-goles-del-empate-entre-tolima-y-nacional-60548
[WinSportsVideo] observa-aqui-los-goles-del-empate-entre-tolima-y-nacional: Downloading webpage
[debug] [MediaStream] Extracting URL: https://mdstrm.com/embed/62dcb875ef12a5526790b552
[MediaStream] 62dcb875ef12a5526790b552: Downloading webpage
[MediaStream] 62dcb875ef12a5526790b552: Downloading m3u8 information
[MediaStream] 62dcb875ef12a5526790b552: Downloading MPD manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 62dcb875ef12a5526790b552:
ID       EXT RESOLUTION FPS │   TBR PROTO │ VCODEC        VBR ACODEC     ABR ASR MORE INFO
─────────────────────────────────────────────────────────────────────────────────────────────────────
f1-a1-x3 m4a audio only     │   94k dash  │ audio only        mp4a.40.2  94k 44k DASH audio, m4a_dash
1        mp4 unknown        │       https │ unknown           unknown
797      mp4 426x240        │  797k m3u8  │ avc1.4d0015  797k mp4a.40.5   0k
f1-v1-x3 mp4 426x240     30 │  797k dash  │ avc1.4d4015  797k video only         DASH video, mp4_dash
1199     mp4 640x360        │ 1199k m3u8  │ avc1.4d001e 1199k mp4a.40.5   0k
f2-v1-x3 mp4 640x360     30 │ 1199k dash  │ avc1.4d401e 1199k video only         DASH video, mp4_dash
1802     mp4 854x480        │ 1802k m3u8  │ avc1.4d001f 1802k mp4a.40.5   0k
f3-v1-x3 mp4 854x480     30 │ 1802k dash  │ avc1.4d401f 1802k video only         DASH video, mp4_dash
3924     mp4 1280x720       │ 3924k m3u8  │ avc1.4d0020 3924k mp4a.40.5   0k
f4-v1-x3 mp4 1280x720    60 │ 3924k dash  │ avc1.4d4020 3924k video only         DASH video, mp4_dash
```

### Embed extraction

```bash
[debug] Command-line config: ['https://www.multimedios.com/television/clases-de-llaves-y-castigos-quien-sabe-mas', '--verbose', '-F']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.11.11 [8b644025b] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: d7b17cbb7
[debug] Python 3.10.7 (CPython x86_64 64bit) - Linux-5.19.0-23-generic-x86_64-with-glibc2.36 (OpenSSL 3.0.5 5 Jul 2022, glibc 2.36)
[debug] exe versions: ffmpeg N-108931-g4dda3b1653-20221104 (setts), ffprobe N-108931-g4dda3b1653-20221104, rtmpdump 2.4
[debug] Optional libraries: Cryptodome-3.15.0, certifi-2022.09.24, mutagen-1.46.0, sqlite3-2.6.0, websockets-10.4
[debug] Proxy map: {}
[debug] Loaded 1730 extractors
[debug] [generic] Extracting URL: https://www.multimedios.com/television/clases-de-llaves-y-castigos-quien-sabe-mas
[generic] clases-de-llaves-y-castigos-quien-sabe-mas: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] clases-de-llaves-y-castigos-quien-sabe-mas: Extracting information
[debug] Looking for embeds
[debug] Identified a MediaStream embed
[debug] [MediaStream] Extracting URL: https://mdstrm.com/embed/63731bab8ec9b308a2c9ed28
[MediaStream] 63731bab8ec9b308a2c9ed28: Downloading webpage
[MediaStream] 63731bab8ec9b308a2c9ed28: Downloading m3u8 information
[MediaStream] 63731bab8ec9b308a2c9ed28: Downloading MPD manifest
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id
[info] Available formats for 63731bab8ec9b308a2c9ed28:
ID       EXT RESOLUTION │   TBR PROTO │ VCODEC        VBR ACODEC     ABR ASR MORE INFO
─────────────────────────────────────────────────────────────────────────────────────────────────
f1-a1-x3 m4a audio only │   94k dash  │ audio only        mp4a.40.2  94k 44k DASH audio, m4a_dash
1        mp4 unknown    │       https │ unknown           unknown
f1-v1-x3 mp4 854x480    │  700k dash  │ avc1.4d401f  700k video only         DASH video, mp4_dash
700      mp4 854x480    │  700k m3u8  │ avc1.4d001f  700k mp4a.40.5   0k
f2-v1-x3 mp4 1280x720   │ 2000k dash  │ avc1.4d401f 2000k video only         DASH video, mp4_dash
2000     mp4 1280x720   │ 2000k m3u8  │ avc1.4d001f 2000k mp4a.40.5   0k
```

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/mediastream.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    }]
+
+    @classmethod
+    def _extract_embed_urls(cls, url, webpage):
+        for mobj in re.finditer(r'<script[^>]+>[^>]*playerMdStream.mdstreamVideo\(\s*[\'"](?P<video_id>\w+)', webpage):
+            yield f'https://mdstrm.com/embed/{mobj.group("video_id")}'
+
+        for mobj in re.finditer(r'<iframe[^>]src\s*=\s*"https://mdstrm.com/(?P<video_type>[\w-]+)/(?P<video_id>\w+)',
+                                webpage):
+            yield f'https://mdstrm.com/{mobj.group("video_type")}/{mobj.group("video_id")}'
```
Comment (pukkandan) [inline yt_dlp/extractor/mediastream.py:72]:
```suggestion
        yield from re.findall(
            r'<iframe[^>]src\s*=\s*"(https://mdstrm.com/[\w-]+/\w+)', webpage)
```

File: yt_dlp/extractor/mediastream.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                class\s*=\s*"[^"]*MediaStreamVideoPlayer[^"]*"[^>]+
+                data-video-id\s*=\s*"(?P<video_id>\w+)\s*"
+                (?:\s*data-video-type\s*=\s*"(?P<video_type>[^"]+))?
+                ''', webpage):
+
+            video_type = 'embed'
+            if mobj.group('video_type') == 'live':
+                video_type = 'live-stream'
+
+            yield f'https://mdstrm.com/{video_type}/{mobj.group("video_id")}'
```
Comment (pukkandan) [inline yt_dlp/extractor/mediastream.py:86]:
```suggestion
            video_type = 'live-stream' if mobj.group('video_type') == 'live else 'embed'
            yield f'https://mdstrm.com/{video_type}/{mobj.group("video_id")}'
```

File: yt_dlp/extractor/mediastream.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                fmts, subs = self._extract_m3u8_formats_and_subtitles(player_config['src'][video_format], video_id)
+                formats.extend(fmts)
+                self._merge_subtitles(subs, target=subtitles)
+            elif video_format == 'mpd':
+                fmts, subs = self._extract_mpd_formats_and_subtitles(player_config['src'][video_format], video_id)
+                formats.extend(fmts)
+                self._merge_subtitles(subs, target=subtitles)
+            else:
+                formats.append({
+                    'url': player_config['src'][video_format],
```
Comment (pukkandan) [inline yt_dlp/extractor/mediastream.py:109]: No info about the format is available?

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/mediastream.py (hunk 3 above)
Comment (elyse0) [reply yt_dlp/extractor/mediastream.py:105]: Nop, just urls

```json
{
   "id":"637307669609130f74cd3a6e",
   "metadata":{
      "custom":"None",
      "accountID":"5b8ea6f89ff52d0770a144c4",
      "customerID":"None",
      "distributor":"None",
   },
   ...
   "src":{
      "hls":"https://mdstrm.com/video/637307669609130f74cd3a6e.m3u8",
      "mpd":"https://mdstrm.com/video/637307669609130f74cd3a6e.mpd",
      "mp4":"https://mdstrm.com/video/637307669609130f74cd3a6e.mp4"
   },
   "type":"recorded",
   "content_type":"video",
   "player_skin":"solidOne",
   "duration":295,
   ...
}
```

**Pull request merged**